### PR TITLE
CBC repeated IV

### PR DESCRIPTION
**Pull request by AleX04Nov**
This PR adds the `CBC repeated IV` demo.

### CODE REVIEW

File: cbc-repeated-iv/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+# need to use aes
+aes = "0.8.1"
+
+# to return structure `anyhow::Result<()>`
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/Cargo.toml]: Cool, but there is no need of documenting every dependency, at least I haven't seen it in the wild :)
It can be useful, you want to specify something extraordinary, like a `TODO` or `FIXME`.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+            let mut line = String::new();
+            println!(
+                "\n{}",
+                "❱".repeat(40).color(Color::Black).on_color(Color::Yellow)
+            );
+            println!("Input key (c(continue)/q(quit)):");
+            std::io::stdin().read_line(&mut line)?;
+            // remove new line char if present and leading and tracing spaces
+            line = line.trim().to_string();
+            match line.to_lowercase().as_str() {
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: These two lines contain two allocations: first in `to_string`, and second in `to_lowercase`.  It could be eliminated to only one:
```suggestion
            match line.trim().to_lowercase().as_str() {
```

File: cbc-repeated-iv/src/cbc_game.rs [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+        }
+
+        Ok(())
+    }
+
+    fn reinitialize_enc_dec(&mut self) {
+        let mut rng = rand::thread_rng();
+        self.key = rng.gen();
+        self.iv = rng.gen();
+    }
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: This method is used only once and during initialization. One of the rust's philosophy is "if the object exists, it should be already in a valid state". In this case, we can move these two lines into `new` method.

Just as example, imagine this as a piece of real crypto code. Making key initialization as a separate method will create possibilities for users to misuse this object and create weaknesses in systems. Moving key generation into `new` will eliminate such possibilities.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        }
+    }
+
+    fn example_eavesdropped_package(&self, iter: usize) {
+        // Stored strings with description for every iteration.
+        let (
+            example_desc_start,
+            example_desc_difference,
+            ascii_art
+        ) = match iter {
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: I would refactor this match statement into separate function, because the `example_eavesdropped_package` is huge and hard to read

File: cbc-repeated-iv/src/cbc_game.rs [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    fn encrypt_text(&self, plaintext_vec: &[u8]) -> Vec<u8> {
+        Aes128CbcEnc::new(&self.key.into(), &self.iv.into())
+            .encrypt_padded_vec_mut::<Pkcs7>(plaintext_vec)
+    }
+
+    fn decrypt_text(&self, ciphertext_vec: &[u8]) -> Vec<u8> {
+        Aes128CbcDec::new(&self.key.into(), &self.iv.into())
+            .decrypt_padded_vec_mut::<Pkcs7>(ciphertext_vec)
+            .ok()
+            .unwrap()
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: Why do you use `unwrap` here? Decryption is operation that by definition should fail in most of the cases. If you are sure that there is no way it could fail, explain it with comments or `expect`. Or just return `anyhow::Result<Vec<u8>>` and propagate error with `?` operator.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+   ▐░▌       ▐░▌▐░▌          ▐░▌       ▐░▌▐░▌          ▐░▌▐░▌       ▐░░░░░░░░░░░░░░░▌▐░█▀▀▀▀▀▀▀▀▀
+   ▐░█▄▄▄▄▄▄▄█░▌▐░█▄▄▄▄▄▄▄▄▄ ▐░█▄▄▄▄▄▄▄█░▌▐░█▄▄▄▄▄▄▄▄▄ ▐░▌ ▐░▌       ▀█░█▀▀▀▀▀▀▀█░█▀ ▐░█▄▄▄▄▄▄▄▄▄
+   ▐░░░░░░░░░░▌ ▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌▐░▌  ▐░▌       ▐░▌       ▐░▌  ▐░░░░░░░░░░░▌
+    ▀▀▀▀▀▀▀▀▀▀   ▀▀▀▀▀▀▀▀▀▀▀  ▀▀▀▀▀▀▀▀▀▀▀  ▀▀▀▀▀▀▀▀▀▀▀  ▀    ▀         ▀         ▀    ▀▀▀▀▀▀▀▀▀▀▀"
+            ),
+            3 => (
+                "Example with package that has the different name field being intercepted",
+                "Block #1 differs. We can assume that data in that block has changed. Also keeping \
+                in mind, that the first block is the same, so structure of json can be the same. \
+                That's why we can guess that usertype has changed and maybe blocks after that...",
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: As I see, almost all of this comments are the same and differ only in the block number. Maybe we can reuse them with fomat strings?

File: cbc-repeated-iv/src/cbc_game.rs [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+    // Same as format_blocks_utf8, but created for formatting CTs
+    // It converts input_vec to hex-encoded string
+    fn format_blocks_hex(input_vec: &[u8], block_size: usize) -> (String, String) {
+        let mut formatted_blocks = String::new();
+        let mut formatted_input = String::new();
+
+        for (block_num, block) in input_vec.chunks(block_size).enumerate() {
+            write!(
+                formatted_blocks,
+                "|            Block #{}            ",
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs:553]: You allow passing any block_size into a function, but headers are precalculated to use 16-byte headers :)
In this case, I would just create a global `BLOCK_SIZE` constant and use it everywhere and together with these headers.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+        println!("{}", self.cbc_same_iv_ascii_art.on_color(Color::Green));
+
+        self.reinitialize_enc_dec();
+
+        self.encrypt_all_texts();
+
+        self.print_known_texts();
+
+        while iter < self.ciphertexts_vec.len() {
+            let mut line = String::new();
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: You could move this line out of the loop, and clear the buffer on every iteration. Just as a fancy and unnecessary optimization :)

File: cbc-repeated-iv/src/cbc_game.rs [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+            CBCGame::format_blocks_hex(&self.known_ciphertext, 16);
+        let formatted_eavesdropped_ct = CBCGame::format_blocks_hex(eavesdropped_packet, 16).1;
+        let formatted_decrypted_eavesdrop = CBCGame::format_blocks_utf8(decrypted_eavesdrop, 16).1;
+
+        println!("{}", ascii_art.on_color(Color::Green));
+        println!("\n{example_desc_start}:");
+        println!("Leaked plaintext:");
+        println!("{formatted_blocks_pt}");
+        println!("{formatted_known_pt}");
+        println!("{}", " ".repeat(145).on_color(Color::Magenta));
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: Where did this number come from?

**Review by G1gg1L3s — COMMENTED** [review]
Also, very important comment about PRs and readme:
The first thing in the PR is its description. Usually it contains a description of the changes, the goal of the PR (why it exists in the first place), technical details and how to run/test it. The exact content depends on a kind of PR, project and people you work with (there are people that will not look at the PR entirely, if there is no description), but it would be good to see at least some of these sections.

In our case, the same is true for readme as well. It's a demo, so we should explain at least how to run it and what it shows.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+        let (formatted_blocks_ct, formatted_known_ct) =
+            CBCGame::format_blocks_hex(&self.known_ciphertext, 16);
+
+        println!("Getting corresponding known CT to leaked PT:");
+        println!("Here's leaked plaintext:");
+        println!("{formatted_blocks_pt}");
+        println!("{formatted_known_pt}\n");
+        println!("Here's corresponding CT to leaked PT:");
+        println!("{formatted_blocks_ct}");
+        println!("{formatted_known_ct}\n\n");
```
Comment (vixentael) [inline cbc-repeated-iv/src/cbc_game.rs]: Let's not use "CT" and "PT" in prints, let's use "ciphertext" and "plaintext" instead.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+
+        println!("Let's wait and eavesdrop other CTs:");
```
Comment (vixentael) [inline cbc-repeated-iv/src/cbc_game.rs]: same

**Comment by vixentael** [comment]
+1 to readme.

i run the example, but what it does? i put "c" 4 times and the program exits. i see wild things in my console, so some descriptions would be nice.

File: cbc-repeated-iv/src/cbc_game.rs (hunk 6 above)
Comment (AleX04Nov) [reply cbc-repeated-iv/src/cbc_game.rs]: Well, the first and the last one iteration have different comments. So I don't think that we need to make our code more complex for the sake of 2 iterations.

File: cbc-repeated-iv/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+You will see a console input.
+
+## CBC Repeated IV
+
+It shows, how reusing init vector can cause big problems. For example - replay attacks, or gaining knowledge from new CipherTexts.
+
+Every iteration show how attacker can gain useful information from new CipherTexts.
+
+Every launch of this demo generates unique keys and init vectors, but due to reusing init vector, this program shows exploitation of reusing init vector.
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/README.md]: The _reusing init vector_ is repeated twice at the end. I would rephrase it to something like:
> Every launch of this demo generates unique key and init vector. But since the IV is reused every iteration, we can gain some information about the content of the messages.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 13]
```diff
@@ -0,0 +1,528 @@
...
+
+    // Creating beautified ascii table
+    // User must send strings (lines), length of each line and padding
+    // We need to send length of each line explicitly, because ColoredString adding extra characters
+    // That are hidden in standard console output
+    fn create_table(
+        strings_arr: &[&String],
+        strings_length: &[usize],
+        padding: usize,
+    ) -> anyhow::Result<String> {
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: This function returns result, but you use only `expect` here. It would be easier and better to use the `?` operator.

The `unwrap` and `expect` are generally used in places, where errors should not occur, but if the do, this is a bug, so we allow process to fail with ugly message and stacktrace.

But the errors that are part of a normal flow, like creating a file, generally are returned to the caller. In this way, a user (of a library for example) can use that error to print a nice message and recover the normal flow.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 14]
```diff
@@ -0,0 +1,528 @@
...
+            ],
+            &[
+                16,
+                String::from_utf8_lossy(&self.known_plaintext).len() + 19,
+                hex::encode(&self.known_ciphertext).len() + 20,
+                hex::encode(eavesdropped_packet).len() + 27,
+            ],
+            3,
+        )
+        .expect("Ascii Table Creation Exception");
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: The same is here. The panic are not exceptions as in other languages. They are (with some omission) unrecoverable.

But if you want just to stop and exit from function in case of some error, just use `?`. It's really easy and convenient.

You can also look at the source code of some rust crates or even standard library, to get a feeling of error handling :)

File: cbc-repeated-iv/src/cbc_game.rs [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+                &format!(
+                    "{}  - eavesdropped CipherText",
+                    hex::encode(eavesdropped_packet)
+                ),
+            ],
+            &[
+                16,
+                String::from_utf8_lossy(&self.known_plaintext).len() + 19,
+                hex::encode(&self.known_ciphertext).len() + 20,
+                hex::encode(eavesdropped_packet).len() + 27,
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: Hard-coded constant are generally a bad sign. In this small demo they could be okay-ish. But, I would suggest you  to look and try out at some ascii table libraries. They, as I remember, support colors, utf8 chars (even emojis) and are generally easy to use.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 16]
```diff
@@ -0,0 +1,528 @@
...
+        }
+    }
+
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        let mut iter = 0;
+        let mut line = String::new();
+
+        println!("{}", self.cbc_same_iv_ascii_art.on_color(Color::Green));
+
+        self.reinitialize_enc_dec();
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs:77]: We already initialize the key and iv. Seems like a leftover :)

File: cbc-repeated-iv/src/cbc_game.rs [hunk 17]
```diff
@@ -0,0 +1,528 @@
...
+        let mut result_string = String::new();
+        let longest_string = strings_length
+            .iter()
+            .max()
+            .expect("Can not retrieve max value from strings_length array");
+        let size = longest_string + padding * 2;
+
+        writeln!(result_string, "╔{}╗", "═".repeat(size)).expect("");
+        writeln!(result_string, "║{}║", " ".repeat(size)).expect("");
+        for string_index in 0..strings_arr.len() {
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: You could replace index-based loop with an iterator-based one
```suggestion
        for (string, length) in string_arr.iter().zip(strings_length) {
```

File: cbc-repeated-iv/src/cbc_game.rs (hunk 16 above)
Comment (AleX04Nov) [reply cbc-repeated-iv/src/cbc_game.rs:77]: I think, that we can leave it. Because it would show that when the interactive-game restarts - new keys are being generated. It's some kind of consistency, so the user can call the `start` method one more time and will remain with new keys and new iv.

Comment (G1gg1L3s) [reply cbc-repeated-iv/src/cbc_game.rs:77]: If you want, sure it can. But in this case I think this method should call `encrypt_all_texts`, so the object will be in consistent state.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 18]
```diff
@@ -0,0 +1,575 @@
...
+        let title_table_format = FormatBuilder::new()
+            .column_separator('║')
+            .borders('║')
+            .separators(&[LinePosition::Top], LineSeparator::new('═', '╦', '╔', '╗'))
+            .separators(
+                &[LinePosition::Intern],
+                LineSeparator::new('═', '╬', '╠', '╣'),
+            )
+            .padding(padding, padding)
+            .build();
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs:456]: What is the difference between these formats? They look almost the same :eyes:

File: cbc-repeated-iv/src/cbc_game.rs [hunk 19]
```diff
@@ -0,0 +1,575 @@
...
+        }
+    }
+
+    // Creating beautified ascii table
+    // User must send strings (lines), length of each line and padding
+    // We need to use old (3 years ago last commit) crate, because new one
+    // Can't handle Colorized strings properly
+    fn create_table(
+        title: &str,
+        data: Vec<Vec<&String>>,
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: Very interesting argument type. Usually it would be written in one of the following ways:
```rust
data: Vec<Vec<String>> // When we want to took an ownership of the entire data structure to modify it and do something with the result (for example print). We took it by value so we can avoid allocations.

data: &[Vec<&str>] // We want just immutable reference to see the content and maybe print or seraillize
```

In our case, any of them will do the trick. And of course, your variant is ok, it just uses `&String` which requires two indirections instead of one, as in `&str`.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 20]
```diff
@@ -0,0 +1,575 @@
...
+        // Get main table as string variable
+        let mut main_table_str = main_table.to_string();
+        // And remove 'new line char' at the end of the string
+        main_table_str.pop();
+
+        // If there is no title => return only main table
+        if title.is_empty() {
+            Ok(main_table_str)
+        }
+        // Otherwise we need to pair them
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs:489]: Just a curiosity: have you tried to use recursive tables, (the example is at the end of [this](https://github.com/phsym/prettytable-rs#using-macros) section of pretty-table.rs)?  Maybe with some style magic it can be done easier. This is just an assumption :)

Comment (AleX04Nov) [reply cbc-repeated-iv/src/cbc_game.rs:489]: No, I think I would be left with more headache and problems than with current logic.

File: cbc-repeated-iv/src/cbc_game.rs (hunk 18 above)
Comment (AleX04Nov) [reply cbc-repeated-iv/src/cbc_game.rs:456]: Yes, almost the same, except the title format does not have a bottom border.
I thought that it would be better to create a format for the title and then expand it with the bottom border for the main table. But style builder actually has method `build`, which I think - does not allow us to expand style. Also, `build` returns `TableFormat` instead of `TableFormatBuilder`.

But I think that I can save the builder into a mutable variable. And build those styles only when I would call `set_format` for my tables.

File: cbc-repeated-iv/src/cbc_game.rs [hunk 21]
```diff
@@ -209,7 +209,7 @@ impl CBCGame {
 
         let ascii_table_step1 = CBCGame::create_table(
             "Step 1:",
-            vec![vec![&format!(
+            &[vec![&format!(
                 "\n{}\n{}\n{}\n{}\n{}\n\n",
                 &"Let's compare known CipherText with the leaked one:".to_string(),
```
Comment (G1gg1L3s) [inline cbc-repeated-iv/src/cbc_game.rs]: Forgot to mention, that this `to_string()` is not needed

**Review by G1gg1L3s — APPROVED** [review]
Okay, looks good

**Review by G1gg1L3s — APPROVED** [review]
Noice

**Review by vixentael — APPROVED** [review]
